File: java/src/main/java/BinaryTree/SubStructure_26.java
package BinaryTree;

/**
 * 输入两棵二叉树A和B，判断B是不是A的子结构。(约定空树不是任意一个树的子结构)
 *
 * B是A的子结构， 即 A中有出现和B相同的结构和节点值。
 */
public class SubStructure_26 {
    // 注意B是A的子结构，B可以提前走到叶子节点，需要注意判断条件
    private boolean same(TreeNode A, TreeNode B){
        if(B==null){
            return true;
        }
        if(A==null){
            return false;
        }
        if(A.val == B.val){
            return same(A.left, B.left) && same(A.right, B.right);
        }else{
            return false;
        }
    }
    private boolean recursionUtil(TreeNode A, TreeNode B){
        if(A == B && A == null){
            return true;
        }
        if(A==null || B==null){
            return false;
        }
        boolean root_res = false;
        if(A.val == B.val){
            root_res = same(A, B);
        }
        if(root_res)
            return true;
        return recursionUtil(A.left, B) || recursionUtil(A.right, B);
    }
    public boolean isSubStructure(TreeNode A, TreeNode B) {
        if(B==null){
            return false;
        }
        return recursionUtil(A, B);
    }
}

File: python/source/cate_bst/sub_structure_26.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from cate_bst.TreeNode import TreeNode
from cate_bst.build import Solution as BuildUtil
class Solution:
    def find(self, root, val):
        if not root:
            return None
        if root.val == val:
            self.find_res.append(root)
        self.find(root.left,val)
        self.find(root.right,val)

    def recur(self, A, B):
        if A == B == None:
            return True
        elif not A and  B:
            return False
        elif A and not B:
            return True
        if A.val == B.val:
            return self.recur(A.left, B.left) and self.recur(A.right, B.right)
        return False

    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        self.find_res = []
        if not B:
            return False
        self.find(A, B.val)
        for node in self.find_res:
            if self.recur(node, B):
                return True
        return False

if __name__ == '__main__':
    util = BuildUtil()
    treeA = util.build_from_array([4,2,3,4,5,6,7,8,9])
    treeB = util.build_from_array([4,8,9])
    so = Solution()
    res = so.isSubStructure(treeA, treeB)
    print(res)